onnxruntime/core/providers/cuda/tensor/gather_elements.cc

```diff
@@ -0,0 +1,136 @@
+      return Status::OK();                                                                                                \
+    }                                                                                                                     \
+    if (Tin_type == DataTypeImpl::GetType<int64_t>()) {                                                                   \
+      const int64_t* indices_data = indices_tensor->template Data<int64_t>();                                             \
+      GatherElementsImpl(                                                                                                 \
+          input_rank,                                                                                                     \
+          reinterpret_cast<const ToCudaType<T>::MappedType*>(input_data),                                                 \
+          input_size,                                                                                                     \
+          input_dims[axis],                                                                                               \
+          gpu_input_strides.GpuPtr(),                                                                                     \
+          indices_data,                                                                                                   \
+          indices_size,                                                                                                   \
+          fdm_indices_strides.GpuPtr(),                                                                                   \
+          axis,                                                                                                           \
+          reinterpret_cast<ToCudaType<T>::MappedType*>(output_data));                                                     \
+      return Status::OK();                                                                                                \
+    }                                                                                                                     \
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "GatherElements op: Type of 'indices' must be int32 or int64"); \
+  }
+
+Status GatherElements::ComputeInternal(OpKernelContext* context) const {
+  // Process input data tensor
+  const auto* input_tensor = context->Input<Tensor>(0);
+  const auto& input_shape = input_tensor->Shape();
+  const auto& input_dims = input_shape.GetDims();
+  const int64_t input_rank = static_cast<int64_t>(input_dims.size());
+  const int64_t input_size = input_shape.Size();
+  if (input_rank < 1)
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "GatherElements op: Cannot operate on scalar input");
+
+  const int axis = static_cast<int>(HandleNegativeAxis(axis_, input_rank));
```

Comment:
Can these initial validations use common code so they're shared between the CPU and GPU versions? e.g. the Slice operator implementations share the setup code

Reply:
Ok.

Reply:
Fixed.